src/neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -50,45 +54,53 @@ protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 acco
-                uint iend = end % Blockchain.DecrementInterval;
-                if (uend >= Blockchain.GenerationAmount.Length)
-                {
-                    uend = (uint)Blockchain.GenerationAmount.Length;
-                    iend = 0;
-                }
-                if (iend == 0)
-                {
-                    uend--;
-                    iend = Blockchain.DecrementInterval;
-                }
-                while (ustart < uend)
-                {
-                    amount += (Blockchain.DecrementInterval - istart) * Blockchain.GenerationAmount[ustart];
-                    ustart++;
-                    istart = 0;
-                }
-                amount += (iend - istart) * Blockchain.GenerationAmount[ustart];
-            }
-            return value * amount * GAS.Factor / TotalAmount;
+            var endRewardItem = snapshot.Storages.TryGet(CreateStorageKey(Prefix_HolderRewardPerBlock).Add(uint.MaxValue - end - 1));
+            var startRewardItem = snapshot.Storages.TryGet(CreateStorageKey(Prefix_HolderRewardPerBlock).Add(uint.MaxValue - start - 1));
+            BigInteger startReward = startRewardItem is null ? 0 : new BigInteger(startRewardItem.Value);
+            BigInteger endReward = endRewardItem is null ? 0 : new BigInteger(endRewardItem.Value);
+            return value * (endReward - startReward) / TotalAmount;
+        }
+
+        private void RecordCumulativeGasByBlock(ApplicationEngine engine)
+        {
+            var gasPerBlock = GetGasPerBlock(engine.Snapshot);
+            RewardRatio rewardRatio = GetRewardRatio(engine.Snapshot);
+            var holderRewards = gasPerBlock * rewardRatio.NeoHolder / 100; // The final calculation should be divided by the total number of NEO
+
+            // Keep track of incremental gains of neo holders
+
+            var holderKeyLeft = CreateStorageKey(Prefix_HolderRewardPerBlock).Add(uint.MaxValue - engine.Snapshot.PersistingBlock.Index - 1);
+            var holderKeyRight = CreateStorageKey(Prefix_HolderRewardPerBlock).Add(uint.MaxValue);
+            var enumerator = engine.Snapshot.Storages.FindRange(holderKeyLeft, holderKeyRight).GetEnumerator();
+            if (enumerator.MoveNext())
+                holderRewards += new BigInteger(enumerator.Current.Value.Value);
+            engine.Snapshot.Storages.Add(holderKeyLeft, new StorageItem() { Value = holderRewards.ToByteArray() });
         }
 
         internal override void Initialize(ApplicationEngine engine)
         {
+            // Initialize economic parameters
+
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_GasPerBlock), new StorageItem
+            {
+                Value = (5 * GAS.Factor).ToByteArray()
+            });
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RewardRatio), new StorageItem(new RewardRatio
+            {
+                NeoHolder = 10,
+                Committee = 5,
+                Voter = 85
```

Comment:
These values should be constants and cannot be modified.

Reply:
Can't committee modify them?

Reply:
No. They must be hard coded.

Reply:
Done.